File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.cpp
class Solution {
public:
    long long maxScore(vector<int>& nums1, vector<int>& nums2, int k) {
        long long sum=0,ans=0;
      vector<pair<int,int>>vect;
        for(int i=0;i<nums1.size();i++){
            vect.push_back({nums2[i],nums1[i]});
        }
        sort(vect.rbegin(),vect.rend());
        priority_queue<int>pq;
        for(int i=0;i<k-1;i++){
            sum+=vect[i].second;
            pq.push(-vect[i].second);
        }
        for(int i=k-1;i<nums1.size();i++){
            sum+=vect[i].second;
            pq.push(-vect[i].second);
            ans=max(ans,sum*vect[i].first);
            sum+=pq.top();
            pq.pop();
        }
        return ans;
    }
};

File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.java
class Solution {
    public long maxScore(int[] nums1, int[] nums2, int k) {
        
        int n=nums1.length;
        int [][]vect=new int[n][2];
        for(int i=0;i<nums1.length;i++){
            vect[i]=new int[]{nums2[i],nums1[i]};
        }
        Arrays.sort(vect,(a,b)->b[0]-a[0]);
        PriorityQueue<Integer>pq=new PriorityQueue<>(k,(a,b)->a-b);
        long res=0,sumS=0;
        for(int[]s:vect){
            pq.add(s[1]);
            sumS+=s[1];
            if(pq.size()>k)sumS-=pq.poll();
            if(pq.size()==k)res=Math.max(res,sumS*s[0]);
            
        }
        return res;
        
    }
}